File: project/frontend/src/lib/nodes/definitions/pytorch/batchnorm.ts
/**
 * PyTorch BatchNorm Layer Node Definition
 */

import { NodeDefinition } from '../../base'
import { NodeMetadata, BackendFramework } from '../../contracts'
import { TensorShape, BlockConfig, ConfigField, BlockType } from '../../../types'

export class BatchNormNode extends NodeDefinition {
  readonly metadata: NodeMetadata = {
    type: 'batchnorm',
    label: 'Batch Normalization',
    category: 'basic',
    color: 'var(--color-accent)',
    icon: 'ChartLineUp',
    description: 'Batch normalization layer',
    framework: BackendFramework.PyTorch
  }

  readonly configSchema: ConfigField[] = [
    {
      name: 'eps',
      label: 'Epsilon',
      type: 'number',
      default: 0.00001,
      min: 0,
      description: 'Value added to denominator for numerical stability'
    },
    {
      name: 'momentum',
      label: 'Momentum',
      type: 'number',
      default: 0.1,
      min: 0,
      max: 1,
      description: 'Momentum for running mean and variance'
    },
    {
      name: 'affine',
      label: 'Learnable Parameters',
      type: 'boolean',
      default: true,
      description: 'Enable learnable affine parameters'
    }
  ]

  computeOutputShape(inputShape: TensorShape | undefined, config: BlockConfig): TensorShape | undefined {
    return inputShape
  }

  validateIncomingConnection(
    sourceNodeType: BlockType,
    sourceOutputShape: TensorShape | undefined,
    targetConfig: BlockConfig
  ): string | undefined {
    // Allow connections from input/dataloader without shape validation
    if (sourceNodeType === 'input' || sourceNodeType === 'dataloader') {
      return undefined
    }

    // Empty and custom nodes are flexible
    if (sourceNodeType === 'empty' || sourceNodeType === 'custom') {
      return undefined
    }

    // Validate dimension requirement (2D or 4D)
    return this.validateDimensions(sourceOutputShape, {
      dims: [2, 4],
      description: '(2D for BatchNorm1d or 4D for BatchNorm2d)'
    })
  }
}

File: python/vision_forge/services/nodes/frameworks/tensorflow/nodes/batchnorm.py
"""TensorFlow BatchNormalization Layer Node"""

from typing import Any, Dict, List, Optional

from ....core.types import ConfigField, Framework, NodeMetadata, TensorShape
from ..base import TensorFlowPassthroughNode


class BatchNormNode(TensorFlowPassthroughNode):
    """
    Batch Normalization Layer.

    Normalizes activations of previous layer.
    Uses tf.keras.layers.BatchNormalization.

    Input: any shape
    Output: same shape as input
    """

    @property
    def metadata(self) -> NodeMetadata:
        return NodeMetadata(
            type="batchnorm",
            label="BatchNorm",
            category="basic",
            color="var(--color-green)",
            icon="Normalize",
            description="Batch normalization",
            framework=Framework.TENSORFLOW
        )

    @property
    def config_schema(self) -> List[ConfigField]:
        return [
            ConfigField(
                name="momentum",
                label="Momentum",
                type="number",
                default=0.99,
                min=0.0,
                max=1.0,
                description="Momentum for moving average"
            ),
            ConfigField(
                name="epsilon",
                label="Epsilon",
                type="number",
                default=0.001,
                description="Small constant for numerical stability"
            )
        ]

    def compute_output_shape(
        self,
        input_shape: Optional[TensorShape],
        config: Dict[str, Any]
    ) -> Optional[TensorShape]:
        if not input_shape:
            return None

        return TensorShape(
            dims=input_shape.dims,
            description="Normalized output"
        )

    def validate_incoming_connection(
        self,
        source_node_type: str,
        source_output_shape: Optional[TensorShape],
        target_config: Dict[str, Any]
    ) -> Optional[str]:
        return None  # Accepts any input
